Add script to perform auto archiving

mathdugre opened the pull request:
## Description
<!--- A clear and concise description of what the dataset is. -->
This PR implement a procedure to automatically archive the CONP datasets for download on the portal.
See #600 for more info.

## Checklist
<!--- Task to do for an approval of the pull request -->
- [ ] Add a chron job on the portal VM to run the script periodically.
- [ ] Generate a `GITHUB_ACCESS_TOKEN` for the script. 


## Related issues (optional)
<!--- Link to issues that would be solved with this pull request -->
#600

mathdugre commented:
@cmadjar , could you review this PR when you have some time. Thank you

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+
+        try:
+            logger.info(f"Installing dataset: {dataset}")
+            install_dataset(dataset)
```
what happens here if the dataset was already installed from a previous archive? 

Should there be a command to update the installed dataset, if the dataset was already installed?

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+            if dataset_size <= args.max_size and is_public:
+                logger.info(f"Downloading dataset: {dataset}")
+                get_dataset(dataset)
+                for submodule in get_proper_submodules(dataset):
```
I am guessing here that you are reusing the same function that is used for the tests to identify whether a submodule is a derived dataset or actually part of the dataset?

mathdugre replied:
Yes this reuses the same function.

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+            # TODO implement notification system.
+            # This will alert when a dataset fails the archiving process.
+            logger.exception(
+                f"FAILURE: could not archive dataset: {dataset} to {args.out_dir}.tar.gz\n{e}"
```
is there a descriptions of what went wrong that comes with the Exception when something goes wrong?

mathdugre replied:
`logger.exception` will log the given message and the stack trace from the occurring exception.
e.g. In the above case, the "FAILURE ..." mesage , the exception message and the stack trace will be logged.

cmadjar reviewed (commented):
@mathdugre Very nice work! I did not test the script yet but had a few questions/comments on the code itself. Let me know.

Thanks!

cmadjar commented:
@mathdugre when running the script, I get the following error regardless on whether I execute the script in the root directory of conp-dataset or the scripts subdirectory:

```
(conp-dataset_virtualenv) cecile@cecile-dev:.../conp-dataset$ python scripts/auto_archive.py 
Traceback (most recent call last):
  File "scripts/auto_archive.py", line 15, in <module>
    from scripts.datalad_utils import get_dataset
ModuleNotFoundError: No module named 'scripts'
```

mathdugre commented:
> @mathdugre when running the script, I get the following error regardless on whether I execute the script in the root directory of conp-dataset or the scripts subdirectory:

@cmadjar, You will have to pass the `PYTHONPATH=$PWD` variable before launching the script.
e.g. `PYTHONPATH=$PWD python scripts/auto_archive.py <out_dir>`

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+    datasets = get_all_datasets() if args.all else get_modified_datasets()
+
+    for dataset in datasets:
+        dataset_name = dataset.removeprefix("projects/")
```
@mathdugre  this will not work for earlier versions of Python. That only exists for v3.9... I am running version 3.7 on my sandbox so the script is currently crashing for me.

mathdugre replied:
@cmadjar which Python version should we support for this tool? Currently, for the test suite we are only supporting 3.9+.

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+        # authentication. However the program will most likely fail.
+        logger.critical(e)
+
+    logger.info(f"Retrieving modfied datasets since {since}")
```
```suggestion
    logger.info(f"Retrieving modified datasets since {since}")
```

mathdugre replied:
Solve in https://github.com/CONP-PCNO/conp-dataset/pull/628/commits/f80ad1cfec032b183c1c5ce549581d703487e5ae

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,202 @@
+    try:
+        gh_access_token = os.environ.get("GITHUB_ACCESS_TOKEN", None)
+        if gh_access_token is None:
+            raise OSError("GITHUB_ACCESS_TOKEN is not defined.")
```
I have defined the GITHUB_ACCESS_TOKEN variable but I always get the following error:
```
(py39_conpdataset) cecile@cecile-dev:.../conp-dataset$ PYTHONPATH=$PWD python scripts/auto_archive.py /data-scratch1/test/out_archive/
CONP-Archive - CRITICAL - GITHUB_ACCESS_TOKEN is not defined.
```

It is in my .bashrc and if I type `echo $GITHUB_ACCESS_TOKEN`, the token is indeed properly saved in the variable. What am I doing wrong?

mathdugre replied:
Was the variable `GITHUB_ACCESS_TOKEN` exported after defining it?

cmadjar replied:
aaaaaah, I forgot to add export in front of the variable name in my `.bashrc`... :( 
Thanks!

cmadjar reviewed (changes requested):
@mathdugre I installed python 3.9 via miniconda and was able to move forward a bit but now I am stuck at the GITHUB_ACCESS_TOKEN step. See below.

I also noticed a small typo in the log displayed to the user.

cmadjar commented:
@mathdugre FYI merging master into your branch should resolve the Circle CI error in this PR.

cmadjar reviewed (commented):
@mathdugre It would be handy if we could specify which dataset we want to run the archive for. Let's say an archive needs to be recreated for a given dataset, I don't see any possibility to do so at the moment.

I also think something is going wrong when there are new datasets on master but locally it is not update. I get the following error for the new datasets:
```
2021-06-21 09:52:22,818 - CONP-Archive - INFO - Retrieving modfied datasets since 2021-06-07 16:28:26.376181-04:00
2021-06-21 09:52:25,348 - CONP-Archive - INFO - Installing dataset: projects/BigBrain_MRISIM
2021-06-21 09:52:27,091 - CONP-Archive - ERROR - FAILURE: could not archive dataset: projects/BigBrain_MRISIM to /data-scratch1/test/out_archive/.tar.gz
Installation failed for dataset: projects/BigBrain_MRISIM
Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/BigBrain_MRISIM',
  'refds': None,
  'status': 'impossible'}]
Traceback (most recent call last):
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 48, in install_dataset
    _install_dataset(dataset_path, recursive=recursive)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 33, in wrapper
    raise last_exception
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 27, in wrapper
    return func(*args, **kwargs)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 43, in _install_dataset
    datalad.api.install(path=full_path, recursive=recursive, on_failure="stop")
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 494, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 482, in return_func
    results = list(results)
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 467, in generator_func
    raise IncompleteResultsError(
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/BigBrain_MRISIM',
  'refds': None,
  'status': 'impossible'}]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data-scratch1/test/conp-dataset/scripts/auto_archive.py", line 160, in <module>
    install_dataset(dataset)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 50, in install_dataset
    raise InstallFailed(f"Installation failed for dataset: {dataset_path}\n{e}")
scripts.datalad_utils.InstallFailed: Installation failed for dataset: projects/BigBrain_MRISIM
Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/BigBrain_MRISIM',
  'refds': None,
  'status': 'impossible'}]
2021-06-21 09:52:27,096 - CONP-Archive - INFO - Installing dataset: projects/Synthetic_Animated_Mouse__SAM___University_of_British_Columbia__Datasets_and_3D_models
2021-06-21 09:53:43,655 - CONP-Archive - INFO - SKIPPED: archive not needed for projects/Synthetic_Animated_Mouse__SAM___University_of_British_Columbia__Datasets_and_3D_models
2021-06-21 09:53:43,658 - CONP-Archive - INFO - Installing dataset: projects/spins_travelling_phantoms
2021-06-21 09:53:44,976 - CONP-Archive - ERROR - FAILURE: could not archive dataset: projects/spins_travelling_phantoms to /data-scratch1/test/out_archive/.tar.gz
Installation failed for dataset: projects/spins_travelling_phantoms
Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/spins_travelling_phantoms',
  'refds': None,
  'status': 'impossible'}]
Traceback (most recent call last):
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 48, in install_dataset
    _install_dataset(dataset_path, recursive=recursive)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 33, in wrapper
    raise last_exception
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 27, in wrapper
    return func(*args, **kwargs)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 43, in _install_dataset
    datalad.api.install(path=full_path, recursive=recursive, on_failure="stop")
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 494, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 482, in return_func
    results = list(results)
  File "/home/cecile/miniconda3/envs/py39_conpdataset/lib/python3.9/site-packages/datalad/interface/utils.py", line 467, in generator_func
    raise IncompleteResultsError(
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/spins_travelling_phantoms',
  'refds': None,
  'status': 'impossible'}]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data-scratch1/test/conp-dataset/scripts/auto_archive.py", line 160, in <module>
    install_dataset(dataset)
  File "/data-scratch1/test/conp-dataset/scripts/datalad_utils.py", line 50, in install_dataset
    raise InstallFailed(f"Installation failed for dataset: {dataset_path}\n{e}")
scripts.datalad_utils.InstallFailed: Installation failed for dataset: projects/spins_travelling_phantoms
Command did not complete successfully. 1 failed:
[{'action': 'get',
  'message': 'path does not exist',
  'path': '/data-scratch1/test/conp-dataset/projects/spins_travelling_phantoms',
  'refds': None,
  'status': 'impossible'}]
2021-06-21 09:53:44,987 - CONP-Archive - INFO - Done archiving the datasets.
```

mathdugre commented:
@cmadjar , I added the option to archive specific datasets.
Also, I fixed the issue with new datasets not being found locally. After fetching the new datasets, it will log a warning message if new datasets are not found locally. To create the archive for those datasets, we will have to `git pull` to update the local repo.
We most likely would want to have a cron job to update the repo then launch the archiving script.

cmadjar commented on scripts/auto_archive.py:
```diff
@@ -0,0 +1,224 @@
+        "-d",
+        type=str,
+        nargs="+",
+        help="Restrict the archive to the datasets specified.",
```
should probably specify here that `projects/` need to be in front of the dataset name or that the dataset specified should be the relative path to the dataset?

mathdugre replied:
@cmadjar, I modified the help for the argument to specify that it expect datasets *paths*. I also change the script to allow for either absolute or relative paths.

cmadjar reviewed (changes requested):
@mathdugre I ran into the following issues when testing:
- I first ran the script as follows and got the following error:
```
python scripts/auto_archive.py -d projects/Reusing-Neuro-Data out_dir /data-scratch1/test/out_archive/

usage: auto_archive.py [-h] [--max-size MAX_SIZE] [--all | --dataset DATASET [DATASET ...]] out_dir
auto_archive.py: error: the following arguments are required: out_dir
```
Looks like the `<out_dir>` positional argument needs to be before any option given to the script but in the help above, out_dir is displayed after the options

- I ran into an error when I tried running the following command because of the final slash after the dataset path:
```
PYTHONPATH=$PWD python scripts/auto_archive.py /data-scratch1/test/out_archive/ --dataset projects/1000GenomesProject/

CONP-Archive - WARNING - The following dataset were not found locally: {'projects/1000GenomesProject/'}
CONP-Archive - INFO - Done archiving the datasets.
```
Might be worth removing any slash at the end of the dataset name.

I have successfully created the Reusing-Neuro-Data dataset in my initial test. 👍  I am still running the 1000GenomesProject on my sandbox. Might take a bit more time. I will continue testing the archiving of several datasets but so far it looks promising :).

mathdugre commented:
> Looks like the <out_dir> positional argument needs to be before any option given to the script but in the help above, out_dir is displayed after the options

Good catch! I fixed this issue in my latest commit https://github.com/CONP-PCNO/conp-dataset/pull/628/commits/495b33cc68809b9f54025f3e26c7c3fe3779f32a. The output directory will now be pass with the flag `--out-dir/-o` to prevent the `-d` flag to think it is another dataset.

Thank you for the feedback and please let me know if other changes are needed.

cmadjar reviewed (approved):
@mathdugre just pulled the latest changes and tested. It all works great 👍 . Thank you!

Looks like the PR is not passing the test though. Maybe you just need to merge master into your branch for it to pass the tests since PERFORM has been fixed by your other PR?
